devcontainer creates files as wrong user (VSC-1495)

### OS

Windows

### Operating System version

Windows 11 (but within the provided docker image for blink)

### Visual Studio Code version

1.93.1

### ESP-IDF version

1.8.1

### Python version

3.12.3

### Doctor command output

---------------------------------------------- ESP-IDF Extension for Visual Studio Code report ---------------------------------------------
OS linux x64 5.15.153.1-microsoft-standard-WSL2 
System environment variable IDF_PYTHON_ENV_PATH 
 /opt/esp/python_env/idf5.4_py3.12_env 
System environment variable PATH 
 /vscode/vscode-server/bin/linux-x64/38c31bc77e0dd6ae88a4e9cc93428cc27a56ba40/bin/remote-cli:/opt/esp/idf/components/espcoredump:/opt/esp/idf/components/partition_table:/opt/esp/idf/components/app_update:/opt/esp/tools/xtensa-esp-elf-gdb/14.2_20240403/xtensa-esp-elf-gdb/bin:/opt/esp/tools/riscv32-esp-elf-gdb/14.2_20240403/riscv32-esp-elf-gdb/bin:/opt/esp/tools/xtensa-esp-elf/esp-14.2.0_20240906/xtensa-esp-elf/bin:/opt/esp/tools/riscv32-esp-elf/esp-14.2.0_20240906/riscv32-esp-elf/bin:/opt/esp/tools/esp32ulp-elf/2.38_20240113/esp32ulp-elf/bin:/opt/esp/tools/cmake/3.30.2/bin:/opt/esp/tools/openocd-esp32/v0.12.0-esp32-20240821/openocd-esp32/bin:/opt/esp/tools/qemu-xtensa/esp_develop_9.0.0_20240606/qemu/bin:/opt/esp/tools/qemu-riscv32/esp_develop_9.0.0_20240606/qemu/bin:/opt/esp/python_env/idf5.4_py3.12_env/bin:/opt/esp/idf/tools:/opt/qemu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin 
System environment variable PYTHON 
 undefined 
Visual Studio Code version 1.93.1 
Visual Studio Code language en 
Visual Studio Code shell /usr/bin/bash 
ESP-IDF Extension version 1.8.1 
Workspace folder /workspaces/blink 
---------------------------------------------------- Extension configuration settings ------------------------------------------------------
ESP-ADF Path (idf.espAdfPath) ${env:ADF_PATH}
ESP-IDF Path (idf.espIdfPath) /opt/esp/idf
ESP-MDF Path (idf.espMdfPath) ${env:MDF_PATH}
ESP-Matter Path (idf.espMatterPath) ${env:ESP_MATTER_PATH}
ESP-HomeKit-SDK Path (idf.espHomeKitSdkPath) ${env:HOMEKIT_PATH}
Custom extra paths (idf.customExtraPaths) 
Custom extra vars (idf.customExtraVars)
    OPENOCD_SCRIPTS: d:\Espressif\tools\openocd-esp32\v0.12.0-esp32-20240318/openocd-esp32/share/openocd/scripts
    IDF_CCACHE_ENABLE: 1
    ESP_ROM_ELF_DIR: d:\Espressif\tools\esp-rom-elfs\20240305/
Virtual env Python Path (idf.pythonBinPath) /opt/esp/python_env/idf5.4_py3.12_env/bin/python
Serial port (idf.port) /dev/ttyUSB1
OpenOCD Configs (idf.openOcdConfigs) interface/ftdi/esp32_devkitj_v1.cfg,board/esp32-wrover.cfg
ESP-IDF Tools Path (idf.toolsPath) /opt/esp
Git Path (idf.gitPath) /usr/bin/git
Notification Mode (idf.notificationMode) All
-------------------------------------------------------- Configurations access -------------------------------------------------------------
Access to ESP-ADF Path (idf.espAdfPath) false
Access to ESP-IDF Path (idf.espIdfPath) true
Access to ESP-MDF Path (idf.espMdfPath) false
Access to ESP-Matter Path (idf.espMatterPath) false
Access to ESP-HomeKit Path (idf.espHomeKitSdkPath) false
Access to ESP-IDF Custom extra paths
Access to Virtual env Python Path (idf.pythonBinPath) true
Access to CMake in environment PATH true
Access to Ninja in environment PATH true
Access to ESP-IDF Tools Path (idf.toolsPath) true
-------------------------------------------------------- Configurations has spaces -------------------------------------------------------------
Spaces in system environment Path false
Spaces in ESP-ADF Path (idf.espAdfPath) false
Spaces in ESP-IDF Path (idf.espIdfPath) false
Spaces in ESP-MDF Path (idf.espMdfPath) false
Spaces in ESP-Matter Path (idf.espMatterPath) false
Spaces in ESP-HomeKit-SDK Path (idf.espHomeKitSdkPath) false
Spaces in ESP-IDF Custom extra paths
Spaces in Virtual env Python Path (idf.pythonBinPath) false
Spaces in ESP-IDF Tools Path (idf.toolsPath) false
----------------------------------------------------------- Executables Versions -----------------------------------------------------------
Git version 2.43.0
ESP-IDF version 5.4.0
Python version 3.12.3
Python's pip version 24.2
-------------------------------------------------- Project configuration settings ----------------------------------------------------------
-------------------------------------------------- Python packages in idf.pythonBinPath ----------------------------------------------------
annotated-types version: 0.7.0
argcomplete version: 3.5.0
bitarray version: 2.9.2
bitstring version: 4.2.3
certifi version: 2024.8.30
cffi version: 1.17.1
charset-normalizer version: 3.3.2
click version: 8.1.7
colorama version: 0.4.6
construct version: 2.10.70
cryptography version: 43.0.1
ecdsa version: 0.19.0
esp-coredump version: 1.11.0
esp-idf-kconfig version: 2.3.0
esp-idf-monitor version: 1.5.0
esp_idf_nvs_partition_gen version: 0.1.3
esp-idf-panic-decoder version: 1.2.1
esp-idf-size version: 1.6.0
esptool version: 4.8.0
freertos-gdb version: 1.0.3
idf-component-manager version: 2.0.3
idna version: 3.10
intelhex version: 2.3.0
jsonref version: 1.1.0
markdown-it-py version: 3.0.0
mdurl version: 0.1.2
packaging version: 24.1
pip version: 24.2
psutil version: 6.0.0
pyclang version: 0.5.0
pycparser version: 2.22
pydantic version: 2.9.2
pydantic_core version: 2.23.4
pydantic-settings version: 2.5.2
pyelftools version: 0.31
pygdbmi version: 0.11.0.0
Pygments version: 2.18.0
pyparsing version: 3.1.4
pyserial version: 3.5
python-dotenv version: 1.0.1
PyYAML version: 6.0.2
reedsolo version: 1.7.0
requests version: 2.32.3
requests-file version: 2.1.0
requests-toolbelt version: 1.0.0
rich version: 13.8.1
setuptools version: 71.0.0
six version: 1.16.0
tqdm version: 4.66.5
typing_extensions version: 4.12.2
urllib3 version: 2.2.3
---------------------------------------------------- Check ESP-IDF python requirements.txt -------------------------------------------------
Check ESP-IDF Python packages Python requirements are satisfied.
---------------------------------------------------- Check ESP-IDF debug adapter requirements.txt ------------------------------------------
Check Debug AdapterPython packages Error
---------------------------------------------------- Visual Studio Code launch.json --------------------------------------------------------
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "gdbtarget",
      "request": "attach",
      "name": "Eclipse CDT GDB Adapter"
    },
    {
      "type": "espidf",
      "name": "Launch",
      "request": "launch"
    }
  ]
} 
---------------------------------------------------- Visual Studio Code c_cpp_properties.json ----------------------------------------------
{
    "configurations": [
        {
            "name": "ESP-IDF",
            "compilerPath": "${config:idf.toolsPathWin}\\tools\\xtensa-esp-elf\\esp-13.2.0_20240530\\xtensa-esp-elf\\bin\\xtensa-esp32-elf-gcc.exe",
            "compileCommands": "${config:idf.buildPath}/compile_commands.json",
            "includePath": [
                "${config:idf.espIdfPath}/components/**",
                "${config:idf.espIdfPathWin}/components/**",
                "${workspaceFolder}/**"
            ],
            "browse": {
                "path": [
                    "${config:idf.espIdfPath}/components",
                    "${config:idf.espIdfPathWin}/components",
                    "${workspaceFolder}"
                ],
                "limitSymbolsToIncludedHeaders": true
            }
        }
    ],
    "version": 4
}
 
----------------------------------------------------------- Latest error -----------------------------------------------------------------
{
    "code": 1,
    "killed": false,
    "signal": null,
    "cmd": "/opt/esp/python_env/idf5.4_py3.12_env/bin/python /opt/esp/idf/tools/check_python_dependencies.py -r <HOMEPATH>/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/esp_debug_adapter/requirements.txt"
}
--------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------- Logfile -----------------------------------------------------------------
{"message":"Open On-Chip Debugger v0.12.0-esp32-20240821 (2024-08-21-14:42)\nLicensed under GNU GPL v2\nFor bug reports, read\n\thttp://openocd.org/doc/doxygen/bugs.html\n","stack":"Error: Open On-Chip Debugger v0.12.0-esp32-20240821 (2024-08-21-14:42)\nLicensed under GNU GPL v2\nFor bug reports, read\n\thttp://openocd.org/doc/doxygen/bugs.html\n\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1424988\n\tat ChildProcess.exithandler (node:child_process:414:7)\n\tat ChildProcess.emit (node:events:519:28)\n\tat maybeClose (node:internal/child_process:1105:16)\n\tat Socket.<anonymous> (node:internal/child_process:457:11)\n\tat Socket.emit (node:events:519:28)\n\tat Pipe.<anonymous> (node:net:338:12)","level":"error","timestamp":"2024-09-28T02:59:17.165Z"}
{"message":"Cannot access filePath: ${env:ADF_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:ADF_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1650824\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T03:16:31.614Z"}
{"message":"Cannot access filePath: ${env:MDF_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:MDF_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651002\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T03:16:31.615Z"}
{"message":"Cannot access filePath: ${env:ESP_MATTER_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:ESP_MATTER_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651094\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T03:16:31.615Z"}
{"message":"Cannot access filePath: ${env:HOMEKIT_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:HOMEKIT_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651190\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T03:16:31.615Z"}
{"message":"Cannot access filePath: ${env:ADF_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:ADF_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1650824\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T04:08:53.914Z"}
{"message":"Cannot access filePath: ${env:MDF_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:MDF_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651002\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T04:08:53.914Z"}
{"message":"Cannot access filePath: ${env:ESP_MATTER_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:ESP_MATTER_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651094\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T04:08:53.915Z"}
{"message":"Cannot access filePath: ${env:HOMEKIT_PATH}","stack":"Error: ENOENT: no such file or directory, access '${env:HOMEKIT_PATH}'\n\tat Object.accessSync (node:fs:244:11)\n\tat Se (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1421806)\n\tat /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651190\n\tat ua (/home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1651892)\n\tat process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n\tat async /home/esp/.vscode-server/extensions/espressif.esp-idf-extension-1.8.1/dist/extension.js:2:1868405","level":"error","timestamp":"2024-09-28T04:08:53.915Z"}

--------------------------------------------------------------------------------------------------------------------------------------------


### Extension

Lots of messages, but none are obviously (to me) relevant. Example:

{"message":"Open On-Chip Debugger v0.12.0-esp32-20240318 (2024-03-18-18:26)\r\nLicensed under GNU GPL v2\r\nFor bug reports, read\r\n\thttp://openocd.org/doc/doxygen/bugs.html\r\n","stack":"Error: Open On-Chip Debugger v0.12.0-esp32-20240318 (2024-03-18-18:26)\r\nLicensed under GNU GPL v2\r\nFor bug reports, read\r\n\thttp://openocd.org/doc/doxygen/bugs.html\r\n\n\tat c:\\Users\\Tobe\\.vscode\\extensions\\espressif.esp-idf-extension-1.8.1\\dist\\extension.js:2:1424988\n\tat ChildProcess.exithandler (node:child_process:415:7)\n\tat ChildProcess.emit (node:events:519:28)\n\tat maybeClose (node:internal/child_process:1105:16)\n\tat ChildProcess._handle.onexit (node:internal/child_process:305:5)","level":"error","timestamp":"2024-09-28T02:59:09.670Z"}

### Description

1) Install vs-code, c++ build tools, esp-idf extension, WSL, dev-containers, blink getting started example source code.
2) Open blink in the provided container.
3) > idf.py --set-target esp32
4) A wall of text appears, with a number of 'permission denied' errors. Se below for the first several lines of output.
5) Some troubleshooting later... (I can provide details)
6) When the 'esp' user writes files, the ownership appears to change to 'ubuntu'. The permission errors appear to be file copy operations for which the 'esp' user has read permissions to access the source, and prior to the start of the script, the target file does not exist. After the script fails, the files mentioned do exist, but have user and group set to 'ubuntu'.
7) When creating a file, the os appears to check for 'esp' permissions, but create the file as user 'ubuntu'.
esp@681d5ae3abf2:/workspaces/blink$ whoami
esp
esp@681d5ae3abf2:/workspaces/blink$ ls test
ls: cannot access 'test': No such file or directory
esp@681d5ae3abf2:/workspaces/blink$ touch test
esp@681d5ae3abf2:/workspaces/blink$ ls -l test
-rwxrwxrwx 1 ubuntu ubuntu 0 Sep 28 04:22 test

 Any ideas or advice are appreciated.

### Debug Message

```plain
esp@681d5ae3abf2:/workspaces/blink$ idf.py set-target esp32
Adding "set-target"'s dependency "fullclean" to list of commands with default set of options.
Executing action: fullclean
Build directory '/workspaces/blink/build' not found. Nothing to clean.
Executing action: set-target
Set Target to: esp32, new sdkconfig will be created.
Running cmake in directory /workspaces/blink/build
Executing "cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DPYTHON=/opt/esp/python_env/idf5.4_py3.12_env/bin/python -DESP_PLATFORM=1 -DIDF_TARGET=esp32 -DCCACHE_ENABLE=1 /workspaces/blink"...
-- Found Git: /usr/bin/git (found version "2.43.0")
CMake Error at /opt/esp/idf/tools/cmake/third_party/GetGitRevisionDescription.cmake:78 (configure_file):
  Operation not permitted
Call Stack (most recent call first):
  /opt/esp/idf/tools/cmake/third_party/GetGitRevisionDescription.cmake:93 (get_git_head_revision)
  /opt/esp/idf/tools/cmake/build.cmake:99 (git_describe)
  /opt/esp/idf/tools/cmake/build.cmake:272 (__build_get_idf_git_revision)
  /opt/esp/idf/tools/cmake/idf.cmake:53 (__build_init)
  /opt/esp/idf/tools/cmake/project.cmake:29 (include)
  CMakeLists.txt:5 (include)


-- ccache will be used for faster recompilation
CMake Error at /opt/esp/tools/cmake/3.30.2/share/cmake-3.30/Modules/CMakeDetermineSystem.cmake:218 (configure_file):
  Operation not permitted
Call Stack (most recent call first):
  /opt/esp/idf/tools/cmake/project.cmake:564 (__project)
  CMakeLists.txt:6 (project)
```


### Other Steps to Reproduce

I have reproduced the error on four different machines, all running Windows 11, two with fresh installs of VS code, VS code C++ tools, WSL, and extensions.

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

--- a/docs_espressif/en/additionalfeatures/docker-container.rst
+++ b/docs_espressif/en/additionalfeatures/docker-container.rst
@@ -123,11 +123,11 @@ for a list of USB serial devices.
 .. note::
   this command needs to be used only one time,unless the computer has restarted. **1-1** is the device's bus id ``<BUSID>`` I would like to bind.
 
-5. After binding, please attach the specified device to WSL with this command in the Powershell command prompt.
+5. After binding, please attach the specified device to WSL with this command in the Powershell command prompt. The ``--auto-attach`` parameter allows the device to be visible in the container after unplug and plug.
 
 .. code-block::
 
-  usbipd attach --wsl --busid <BUSID>
+  usbipd attach --wsl --busid <BUSID> --auto-attach
 
 6. At last, let us check if it works well on both side and type this command on WSL side.
 
@@ -230,6 +230,9 @@ Create a Container
 .. image:: ../../../media/tutorials/using_docker_container/create_container.gif
 
 At this moment, you can start to use the ``Blink`` example project for building, flashing, monitoring, debugging, etc.
+
+.. warning::
+  * In order to have access to the serial port from the Docker container, make sure you have attached the device with ``usbipd attach --wsl --busid <BUSID> --auto-attach`` **BEFORE** opening the folder in container in VS Code otherwise it won't be visible. If you want to be able to plug and unplug the device and still see it in the docker container don't forget the  ``--auto-attach`` usbipd parameter.
 
 3. Here taking the esp32-c3 as an example, users only need to change the target device from ``esp32`` to ``esp32-c3``, as below:
 

--- a/l10n/bundle.l10n.es.json
+++ b/l10n/bundle.l10n.es.json
@@ -62,6 +62,8 @@
   "{destFolder} already exists.": "{destFolder} ya existe.",
   "ESP-IDF Project has been imported": "El proyecto ESP-IDF ha sido importado.",
   "Merging binaries for flashing": "Fusionar binarios para flashear",
+  "ESP-IDF: Starting ESP-IDF QEMU Debug": "ESP-IDF: Iniciando la depuración QEMU de ESP-IDF",
+  "Error launching QEMU debugging": "Error al iniciar la depuración de QEMU",
   "Select a partition to use": "Seleccione una partición para usar",
   "Enter custom partition table offset": "Ingrese el desplazamiento de la tabla de particiones personalizada",
   "Flash binary to this partition": "Flash binario a esta partición",

--- a/l10n/bundle.l10n.pt.json
+++ b/l10n/bundle.l10n.pt.json
@@ -62,6 +62,8 @@
   "{destFolder} already exists.": "{destFolder} já existe.",
   "ESP-IDF Project has been imported": "O projeto ESP-IDF foi importado",
   "Merging binaries for flashing": "Mesclando binários para piscar",
+  "ESP-IDF: Starting ESP-IDF QEMU Debug": "ESP-IDF: Iniciando a depuração ESP-IDF QEMU",
+  "Error launching QEMU debugging": "Erro ao iniciar a depuração do QEMU",
   "Select a partition to use": "Selecione uma partição para usar",
   "Enter custom partition table offset": "Insira o deslocamento da tabela de partição personalizada",
   "Flash binary to this partition": "Flash binário para esta partição",

--- a/l10n/bundle.l10n.ru.json
+++ b/l10n/bundle.l10n.ru.json
@@ -62,6 +62,8 @@
   "{destFolder} already exists.": "{destFolder} уже существует.",
   "ESP-IDF Project has been imported": "Проект ESP-IDF импортирован.",
   "Merging binaries for flashing": "Объединение бинарников для перепрошивки",
+  "ESP-IDF: Starting ESP-IDF QEMU Debug": "ESP-IDF: запуск отладки ESP-IDF QEMU",
+  "Error launching QEMU debugging": "Ошибка запуска отладки QEMU",
   "Select a partition to use": "Выберите раздел для использования",
   "Enter custom partition table offset": "Введите пользовательское смещение таблицы разделов",
   "Flash binary to this partition": "Записать двоичный файл в этот раздел",

--- a/l10n/bundle.l10n.zh-CN.json
+++ b/l10n/bundle.l10n.zh-CN.json
@@ -62,6 +62,8 @@
   "{destFolder} already exists.": "{destFolder} 已存在。",
   "ESP-IDF Project has been imported": "ESP-IDF 项目已导入",
   "Merging binaries for flashing": "合并二进制文件以进行闪烁",
+  "ESP-IDF: Starting ESP-IDF QEMU Debug": "ESP-IDF：启动 ESP-IDF QEMU 调试",
+  "Error launching QEMU debugging": "启动 QEMU 调试时出错",
   "Select a partition to use": "选择要使用的分区",
   "Enter custom partition table offset": "输入自定义分区表偏移量",
   "Flash binary to this partition": "将二进制文件烧录到该分区",

--- a/src/build/buildCmd.ts
+++ b/src/build/buildCmd.ts
@@ -86,6 +86,15 @@ export async function buildCommand(
         await TaskManager.runTasks();
       }
     }
+    const buildDirPath = readParameter(
+      "idf.buildPath",
+      workspace
+    ) as string;
+    const qemuBinPath = join(buildDirPath, "merged_qemu.bin");
+    const qemuBinExists = await pathExists(qemuBinPath);
+    if (qemuBinExists) {
+      await vscode.workspace.fs.delete(vscode.Uri.file(qemuBinPath));
+    }
     if (!cancelToken.isCancellationRequested) {
       updateIdfComponentsTree(workspace);
       Logger.infoNotify("Build Successfully");

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -356,6 +356,7 @@ export async function activate(context: vscode.ExtensionContext) {
     } else {
       UpdateCmakeLists.updateSrcsInCmakeLists(e.fsPath, srcOp.other);
     }
+    await binTimestampEventFunc(e);
   });
   context.subscriptions.push(srcWatchOnChangeDisposable);
 
@@ -374,7 +375,7 @@ export async function activate(context: vscode.ExtensionContext) {
     const qemuBinPath = path.join(buildDirPath, "merged_qemu.bin");
     const qemuBinExists = await pathExists(qemuBinPath);
     if (qemuBinExists) {
-      vscode.workspace.fs.delete(vscode.Uri.file(qemuBinPath));
+      await vscode.workspace.fs.delete(vscode.Uri.file(qemuBinPath));
     }
   };
 

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -2468,7 +2468,7 @@ export async function activate(context: vscode.ExtensionContext) {
         {
           cancellable: true,
           location: ProgressLocation,
-          title: "ESP-IDF: Starting ESP-IDF QEMU Debug",
+          title: vscode.l10n.t("ESP-IDF: Starting ESP-IDF QEMU Debug"),
         },
         async (
           progress: vscode.Progress<{ message: string; increment: number }>,
@@ -2523,7 +2523,7 @@ export async function activate(context: vscode.ExtensionContext) {
           } catch (error) {
             const msg = error.message
               ? error.message
-              : "Error launching QEMU debugging";
+              : vscode.l10n.t("Error launching QEMU debugging");
             Logger.errorNotify(msg, error, "extension qemu debug");
           }
         }

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -360,7 +360,7 @@ export async function activate(context: vscode.ExtensionContext) {
   context.subscriptions.push(srcWatchOnChangeDisposable);
 
   const buildWatcher = vscode.workspace.createFileSystemWatcher(
-    ".bin_timestamp",
+    "**/.bin_timestamp",
     true,
     false,
     true

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -110,7 +110,7 @@ import { configureProjectWithGcov } from "./coverage/configureProject";
 import { ComponentManagerUIPanel } from "./component-manager/panel";
 import { verifyAppBinary } from "./espIdf/debugAdapter/verifyApp";
 import { mergeFlashBinaries } from "./qemu/mergeFlashBin";
-import { IQemuOptions, QemuManager } from "./qemu/qemuManager";
+import { QemuLaunchMode, QemuManager } from "./qemu/qemuManager";
 import {
   PartitionItem,
   PartitionTreeDataProvider,
@@ -359,6 +359,26 @@ export async function activate(context: vscode.ExtensionContext) {
   });
   context.subscriptions.push(srcWatchOnChangeDisposable);
 
+  const buildWatcher = vscode.workspace.createFileSystemWatcher(
+    ".bin_timestamp",
+    true,
+    false,
+    true
+  );
+
+  const buildWatcherDisposable = buildWatcher.onDidChange(async (e) => {
+    const buildDirPath = idfConf.readParameter(
+      "idf.buildPath",
+      workspaceRoot
+    ) as string;
+    const qemuBinPath = path.join(buildDirPath, "merged_qemu.bin");
+    const qemuBinExists = await pathExists(qemuBinPath);
+    if (qemuBinExists) {
+      vscode.workspace.fs.delete(vscode.Uri.file(qemuBinPath));
+    }
+  });
+  context.subscriptions.push(buildWatcherDisposable);
+
   vscode.workspace.onDidChangeWorkspaceFolders(async (e) => {
     if (PreCheck.isWorkspaceFolderOpen()) {
       for (const ws of e.removed) {
@@ -426,9 +446,6 @@ export async function activate(context: vscode.ExtensionContext) {
         workspace: workspaceRoot,
       } as IOpenOCDConfig;
       openOCDManager.configureServer(openOCDConfig);
-      qemuManager.configure({
-        workspaceFolder: workspaceRoot,
-      } as IQemuOptions);
     }
     ConfserverProcess.dispose();
   });
@@ -964,9 +981,6 @@ export async function activate(context: vscode.ExtensionContext) {
           workspace: workspaceRoot,
         } as IOpenOCDConfig;
         openOCDManager.configureServer(openOCDConfig);
-        qemuManager.configure({
-          workspaceFolder: workspaceRoot,
-        } as IQemuOptions);
         ConfserverProcess.dispose();
         const coverageOptions = getCoverageOptions(workspaceRoot);
         covRenderer = new CoverageRenderer(workspaceRoot, coverageOptions);
@@ -1173,10 +1187,6 @@ export async function activate(context: vscode.ExtensionContext) {
       }
     } else if (e.affectsConfiguration("idf.espIdfPath" + winFlag)) {
       ESP.URL.Docs.IDF_INDEX = undefined;
-    } else if (e.affectsConfiguration("idf.qemuTcpPort")) {
-      qemuManager.configure({
-        tcpPort: idfConf.readParameter("idf.qemuTcpPort", workspaceRoot),
-      } as IQemuOptions);
     } else if (e.affectsConfiguration("idf.port" + winFlag)) {
       if (statusBarItems && statusBarItems["port"]) {
         statusBarItems["port"].text =
@@ -2453,8 +2463,7 @@ export async function activate(context: vscode.ExtensionContext) {
         qemuManager.stop();
         await utils.sleep(1000);
       }
-      qemuManager.configureWithDefValues();
-      qemuManager.start();
+      qemuManager.start(QemuLaunchMode.Debug, workspaceRoot);
       const gdbPath = await utils.getToolchainPath(workspaceRoot, "gdb");
       const workspaceFolder = vscode.workspace.getWorkspaceFolder(
         workspaceRoot
@@ -3865,22 +3874,21 @@ function createQemuMonitor() {
     if (isQemuLaunched) {
       qemuManager.stop();
     }
+    const buildDirPath = idfConf.readParameter(
+      "idf.buildPath",
+      workspaceRoot
+    ) as string;
+    const qemuBinExists = await pathExists(
+      path.join(buildDirPath, "merged_qemu.bin")
+    );
+    if (!qemuBinExists) {
+      await mergeFlashBinaries(workspaceRoot);
+    }
     const qemuTcpPort = idfConf.readParameter(
       "idf.qemuTcpPort",
       workspaceRoot
-    ) as number;
-    qemuManager.configure({
-      launchArgs: [
-        "-nographic",
-        "-machine",
-        "esp32",
-        "-drive",
-        "file=build/merged_qemu.bin,if=mtd,format=raw",
-        "-monitor stdio",
-        `-serial tcp::${qemuTcpPort},server,nowait`,
-      ],
-    } as IQemuOptions);
-    await qemuManager.start();
+    ) as string;
+    await qemuManager.start(QemuLaunchMode.Monitor, workspaceRoot);
     if (IDFMonitor.terminal) {
       await utils.sleep(1000);
     }

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -2455,51 +2455,79 @@ export async function activate(context: vscode.ExtensionContext) {
 
   registerIDFCommand("espIdf.qemuDebug", () => {
     PreCheck.perform([openFolderCheck], async () => {
-      if (IDFMonitor.terminal) {
-        IDFMonitor.terminal.sendText(ESP.CTRL_RBRACKET);
-      }
-      const buildDirPath = idfConf.readParameter(
-        "idf.buildPath",
+      const notificationMode = idfConf.readParameter(
+        "idf.notificationMode",
         workspaceRoot
       ) as string;
-      const qemuBinExists = await pathExists(
-        path.join(buildDirPath, "merged_qemu.bin")
-      );
-      if (!qemuBinExists) {
-        await mergeFlashBinaries(workspaceRoot);
-      }
-      if (qemuManager.isRunning()) {
-        qemuManager.stop();
-        await utils.sleep(1000);
-      }
-      qemuManager.start(QemuLaunchMode.Debug, workspaceRoot);
-      const gdbPath = await utils.getToolchainPath(workspaceRoot, "gdb");
-      const workspaceFolder = vscode.workspace.getWorkspaceFolder(
-        workspaceRoot
-      );
-      await vscode.debug.startDebugging(workspaceFolder, {
-        name: "GDB QEMU",
-        type: "gdbtarget",
-        request: "attach",
-        sessionID: "qemu.debug.session",
-        gdb: gdbPath,
-        initCommands: [
-          "set remote hardware-watchpoint-limit {IDF_TARGET_CPU_WATCHPOINT_NUM}",
-          "mon reset halt",
-          "maintenance flush register-cache",
-          "thb app_main",
-        ],
-        target: {
-          type: "remote",
-          host: "localhost",
-          port: "1234",
+      const ProgressLocation =
+        notificationMode === idfConf.NotificationMode.All ||
+        notificationMode === idfConf.NotificationMode.Notifications
+          ? vscode.ProgressLocation.Notification
+          : vscode.ProgressLocation.Window;
+      await vscode.window.withProgress(
+        {
+          cancellable: true,
+          location: ProgressLocation,
+          title: "ESP-IDF: Starting ESP-IDF QEMU Debug",
         },
-      });
-      vscode.debug.onDidTerminateDebugSession(async (session) => {
-        if (session.configuration.sessionID === "qemu.debug.session") {
-          qemuManager.stop();
+        async (
+          progress: vscode.Progress<{ message: string; increment: number }>,
+          cancelToken: vscode.CancellationToken
+        ) => {
+          try {
+            if (IDFMonitor.terminal) {
+              IDFMonitor.terminal.sendText(ESP.CTRL_RBRACKET);
+            }
+            const buildDirPath = idfConf.readParameter(
+              "idf.buildPath",
+              workspaceRoot
+            ) as string;
+            const qemuBinExists = await pathExists(
+              path.join(buildDirPath, "merged_qemu.bin")
+            );
+            if (!qemuBinExists) {
+              await mergeFlashBinaries(workspaceRoot);
+            }
+            if (qemuManager.isRunning()) {
+              qemuManager.stop();
+              await utils.sleep(1000);
+            }
+            await qemuManager.start(QemuLaunchMode.Debug, workspaceRoot);
+            const gdbPath = await utils.getToolchainPath(workspaceRoot, "gdb");
+            const workspaceFolder = vscode.workspace.getWorkspaceFolder(
+              workspaceRoot
+            );
+            await vscode.debug.startDebugging(workspaceFolder, {
+              name: "GDB QEMU",
+              type: "gdbtarget",
+              request: "attach",
+              sessionID: "qemu.debug.session",
+              gdb: gdbPath,
+              initCommands: [
+                "set remote hardware-watchpoint-limit {IDF_TARGET_CPU_WATCHPOINT_NUM}",
+                "mon reset halt",
+                "maintenance flush register-cache",
+                "thb app_main",
+              ],
+              target: {
+                type: "remote",
+                host: "localhost",
+                port: "1234",
+              },
+            });
+            vscode.debug.onDidTerminateDebugSession(async (session) => {
+              if (session.configuration.sessionID === "qemu.debug.session") {
+                qemuManager.stop();
+              }
+            });
+          } catch (error) {
+            const msg = error.message
+              ? error.message
+              : "Error launching QEMU debugging";
+            Logger.errorNotify(msg, error, "extension qemu debug");
+          }
         }
-      });
+      );
     });
   });
 
@@ -3879,34 +3907,62 @@ const flash = (
 
 function createQemuMonitor() {
   PreCheck.perform([openFolderCheck], async () => {
-    const isQemuLaunched = qemuManager.isRunning();
-    if (isQemuLaunched) {
-      qemuManager.stop();
-    }
-    const buildDirPath = idfConf.readParameter(
-      "idf.buildPath",
+    const notificationMode = idfConf.readParameter(
+      "idf.notificationMode",
       workspaceRoot
     ) as string;
-    const qemuBinExists = await pathExists(
-      path.join(buildDirPath, "merged_qemu.bin")
+    const ProgressLocation =
+      notificationMode === idfConf.NotificationMode.All ||
+      notificationMode === idfConf.NotificationMode.Notifications
+        ? vscode.ProgressLocation.Notification
+        : vscode.ProgressLocation.Window;
+    await vscode.window.withProgress(
+      {
+        cancellable: true,
+        location: ProgressLocation,
+        title: "ESP-IDF: Starting ESP-IDF QEMU Monitor",
+      },
+      async (
+        progress: vscode.Progress<{ message: string; increment: number }>,
+        cancelToken: vscode.CancellationToken
+      ) => {
+        try {
+          const isQemuLaunched = qemuManager.isRunning();
+          if (isQemuLaunched) {
+            qemuManager.stop();
+          }
+          const buildDirPath = idfConf.readParameter(
+            "idf.buildPath",
+            workspaceRoot
+          ) as string;
+          const qemuBinExists = await pathExists(
+            path.join(buildDirPath, "merged_qemu.bin")
+          );
+          if (!qemuBinExists) {
+            await mergeFlashBinaries(workspaceRoot);
+          }
+          const qemuTcpPort = idfConf.readParameter(
+            "idf.qemuTcpPort",
+            workspaceRoot
+          ) as string;
+          await qemuManager.start(QemuLaunchMode.Monitor, workspaceRoot);
+          if (IDFMonitor.terminal) {
+            await utils.sleep(1000);
+          }
+          const serialPort = `socket://localhost:${qemuTcpPort}`;
+          const noReset = idfConf.readParameter(
+            "idf.monitorNoReset",
+            workspaceRoot
+          ) as boolean;
+          await createNewIdfMonitor(workspaceRoot, noReset, serialPort);
+        } catch (error) {
+          const msg = error.message
+            ? error.message
+            : "Error launching QEMU monitor";
+          Logger.errorNotify(msg, error, "extension qemu monitor");
+        }
+      }
     );
-    if (!qemuBinExists) {
-      await mergeFlashBinaries(workspaceRoot);
-    }
-    const qemuTcpPort = idfConf.readParameter(
-      "idf.qemuTcpPort",
-      workspaceRoot
-    ) as string;
-    await qemuManager.start(QemuLaunchMode.Monitor, workspaceRoot);
-    if (IDFMonitor.terminal) {
-      await utils.sleep(1000);
-    }
-    const serialPort = `socket://localhost:${qemuTcpPort}`;
-    const noReset = idfConf.readParameter(
-      "idf.monitorNoReset",
-      workspaceRoot
-    ) as boolean;
-    await createNewIdfMonitor(workspaceRoot, noReset, serialPort);
   });
 }
 

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -361,12 +361,12 @@ export async function activate(context: vscode.ExtensionContext) {
 
   const buildWatcher = vscode.workspace.createFileSystemWatcher(
     "**/.bin_timestamp",
-    true,
+    false,
     false,
     true
   );
 
-  const buildWatcherDisposable = buildWatcher.onDidChange(async (e) => {
+  const binTimestampEventFunc = async (e: vscode.Uri) => {
     const buildDirPath = idfConf.readParameter(
       "idf.buildPath",
       workspaceRoot
@@ -376,8 +376,16 @@ export async function activate(context: vscode.ExtensionContext) {
     if (qemuBinExists) {
       vscode.workspace.fs.delete(vscode.Uri.file(qemuBinPath));
     }
-  });
+  };
+
+  const buildWatcherDisposable = buildWatcher.onDidChange(
+    binTimestampEventFunc
+  );
   context.subscriptions.push(buildWatcherDisposable);
+  const buildWatcherCreateDisposable = buildWatcher.onDidCreate(
+    binTimestampEventFunc
+  );
+  context.subscriptions.push(buildWatcherCreateDisposable);
 
   vscode.workspace.onDidChangeWorkspaceFolders(async (e) => {
     if (PreCheck.isWorkspaceFolderOpen()) {

--- a/src/qemu/qemuManager.ts
+++ b/src/qemu/qemuManager.ts
@@ -114,7 +114,7 @@ export class QemuManager extends EventEmitter {
         "-machine",
         idfTarget,
         "-drive",
-        `file=${qemuFile.fsPath},if=mtd,format=raw`,
+        `file='${qemuFile.fsPath}',if=mtd,format=raw`,
       ];
     } else {
       return [

--- a/src/qemu/qemuManager.ts
+++ b/src/qemu/qemuManager.ts
@@ -25,22 +25,20 @@ import {
   TreeItemCheckboxState,
   Uri,
   window,
-  workspace,
 } from "vscode";
 import { ESP } from "../config";
 import { readParameter } from "../idfConfiguration";
 import { Logger } from "../logger/logger";
-import { appendIdfAndToolsToPath, isBinInPath, PreCheck } from "../utils";
 import { statusBarItems } from "../statusBar";
 import {
   CommandKeys,
   createCommandDictionary,
 } from "../cmdTreeView/cmdStore";
+import { appendIdfAndToolsToPath, isBinInPath } from "../utils";
 
-export interface IQemuOptions {
-  launchArgs: string[];
-  tcpPort: string;
-  workspaceFolder: Uri;
+export enum QemuLaunchMode {
+  Debug,
+  Monitor,
 }
 
 export class QemuManager extends EventEmitter {
@@ -51,14 +49,12 @@ export class QemuManager extends EventEmitter {
     return QemuManager.instance;
   }
   private static instance: QemuManager;
-  private execString = "qemu-system-xtensa";
   private qemuTerminal: Terminal;
-  private options: IQemuOptions;
   private _statusBarItem: StatusBarItem;
 
   private constructor() {
     super();
-    this.configureWithDefValues();
+    this.registerQemuStatusBarItem();
   }
 
   public statusBarItem(): StatusBarItem {
@@ -71,30 +67,20 @@ export class QemuManager extends EventEmitter {
   }
 
   public async commandHandler() {
-    const pickItems = [];
-    if (!QemuManager.instance.isRunning()) {
-      pickItems.push({
-        label: "Start QEMU",
-        description: "",
-      });
-    } else {
-      pickItems.push({
+    const pickItems = [
+      {
         label: "Stop QEMU",
         description: "",
-      });
-    }
+      },
+    ];
     const selectedOption = await window.showQuickPick(pickItems);
     if (!selectedOption) {
       return;
     }
-
     try {
       switch (selectedOption.label) {
-        case "Start QEMU":
-          await QemuManager.instance.start();
-          break;
         case "Stop QEMU":
-          await QemuManager.instance.stop();
+          QemuManager.instance.stop();
           break;
         default:
           break;
@@ -105,103 +91,91 @@ export class QemuManager extends EventEmitter {
     }
   }
 
-  public configure(config: IQemuOptions) {
-    if (!this.options) {
-      this.options = {} as IQemuOptions;
-    }
-    if (config.launchArgs) {
-      this.options.launchArgs = config.launchArgs;
-    }
-    if (config.tcpPort) {
-      this.options.tcpPort = config.tcpPort;
-    }
-    if (config.workspaceFolder) {
-      this.options.workspaceFolder = config.workspaceFolder;
-    }
-    this.registerQemuStatusBarItem();
-  }
+  public getLaunchArguments(
+    mode: QemuLaunchMode,
+    idfTarget: string,
+    workspaceFolder: Uri
+  ) {
+    const buildPath = readParameter(
+      "idf.buildPath",
+      workspaceFolder
+    ) as string;
+    const qemuFile = Uri.joinPath(Uri.file(buildPath), "merged_qemu.bin");
+    const qemuTcpPort = readParameter(
+      "idf.qemuTcpPort",
+      workspaceFolder
+    ) as string;
 
-  public configureWithDefValues() {
-    let workspaceFolder: Uri;
-    if (PreCheck.isWorkspaceFolderOpen()) {
-      workspaceFolder = workspace.workspaceFolders[0].uri;
-    }
-    const defOptions = {
-      launchArgs: [
+    if (mode === QemuLaunchMode.Debug) {
+      return [
         "-nographic",
         "-s",
         "-S",
         "-machine",
-        "esp32",
+        idfTarget,
         "-drive",
-        "file=build/merged_qemu.bin,if=mtd,format=raw",
-      ],
-      tcpPort: readParameter("idf.qemuTcpPort", workspaceFolder),
-      workspaceFolder,
-    } as IQemuOptions;
-    this.configure(defOptions);
+        `file=${qemuFile.fsPath},if=mtd,format=raw`,
+      ];
+    } else {
+      return [
+        "-nographic",
+        "-machine",
+        idfTarget,
+        "-drive",
+        `file=${qemuFile.fsPath},if=mtd,format=raw`,
+        "-monitor stdio",
+        `-serial tcp::${qemuTcpPort},server,nowait`,
+      ];
+    }
   }
 
   public isRunning(): boolean {
     return !!this.qemuTerminal;
   }
 
-  public async promptToQemuLaunch(): Promise<boolean> {
-    if (QemuManager.instance && QemuManager.instance.isRunning()) {
-      return true;
-    }
-    const launchQemuResponse = await window.showInformationMessage(
-      "QEMU is not running, Do you want to execute it?",
-      { modal: true },
-      { title: "Yes" },
-      { title: "Cancel", isCloseAffordance: true }
-    );
-    if (launchQemuResponse && launchQemuResponse.title === "Yes") {
-      await QemuManager.init().start();
-      return true;
-    }
-    return false;
-  }
-
-  public async start() {
+  public async start(mode: QemuLaunchMode, workspaceFolder: Uri) {
     if (this.isRunning()) {
       return;
     }
     const modifiedEnv = await appendIdfAndToolsToPath(
-      this.options.workspaceFolder
+      workspaceFolder
     );
+    const qemuExecutable =
+      modifiedEnv.IDF_TARGET === "esp32"
+        ? "qemu-system-xtensa"
+        : modifiedEnv.IDF_TARGET === "esp32c3"
+        ? "qemu-system-riscv32"
+        : "";
+    if (!qemuExecutable) {
+      throw new Error(
+        `Current IDF_TARGET ${modifiedEnv.IDF_TARGET} is not supported in Espressif QEMU. Only esp32 or esp32c3`
+      );
+    }
     const isQemuBinInPath = await isBinInPath(
-      this.execString,
-      this.options.workspaceFolder.fsPath,
+      qemuExecutable,
+      workspaceFolder.fsPath,
       modifiedEnv
     );
     if (!isQemuBinInPath) {
-      throw new Error("qemu-system-xtensa is not in PATH or access is denied");
+      throw new Error(
+        `${qemuExecutable} is not found in PATH or access is denied`
+      );
     }
-    if (typeof this.options === "undefined") {
-      throw new Error("No QEMU options found.");
-    }
+
+    const qemuArgs: string[] = this.getLaunchArguments(mode, modifiedEnv.IDF_TARGET, workspaceFolder);
     if (
-      typeof this.options.launchArgs === "undefined" ||
-      this.options.launchArgs.length < 1
+      typeof qemuArgs === "undefined" ||
+      qemuArgs.length < 1
     ) {
       throw new Error("No QEMU launch arguments found.");
     }
-    if (typeof this.options.tcpPort === "undefined") {
-      throw new Error("No QEMU tcp port for serial port was found.");
-    }
-
-    const qemuArgs: string[] = [];
-    this.options.launchArgs.forEach((arg) => {
-      qemuArgs.push(arg);
-    });
 
     if (typeof this.qemuTerminal === "undefined") {
       this.qemuTerminal = window.createTerminal({
         name: "ESP-IDF QEMU",
         env: modifiedEnv,
         cwd:
-          this.options.workspaceFolder.fsPath ||
+          workspaceFolder.fsPath ||
           modifiedEnv.IDF_PATH ||
           process.cwd(),
         shellArgs: [],
@@ -214,7 +188,7 @@ export class QemuManager extends EventEmitter {
         }
       });
     }
-    this.qemuTerminal.sendText(`${this.execString} ${qemuArgs.join(" ")}`);
+    this.qemuTerminal.sendText(`${qemuExecutable} ${qemuArgs.join(" ")}`);
     this.qemuTerminal.show(true);
     this.updateStatusText("❇️ ESP-IDF: QEMU Server (Running)");
   }

--- a/src/qemu/qemuManager.ts
+++ b/src/qemu/qemuManager.ts
@@ -18,6 +18,7 @@
 
 import { EventEmitter } from "events";
 import {
+  commands,
   env,
   StatusBarAlignment,
   StatusBarItem,
@@ -30,10 +31,7 @@ import { ESP } from "../config";
 import { readParameter } from "../idfConfiguration";
 import { Logger } from "../logger/logger";
 import { statusBarItems } from "../statusBar";
-import {
-  CommandKeys,
-  createCommandDictionary,
-} from "../cmdTreeView/cmdStore";
+import { CommandKeys, createCommandDictionary } from "../cmdTreeView/cmdStore";
 import { appendIdfAndToolsToPath, isBinInPath } from "../utils";
 
 export enum QemuLaunchMode {
@@ -69,6 +67,14 @@ export class QemuManager extends EventEmitter {
   public async commandHandler() {
     const pickItems = [
       {
+        label: "QEMU Monitor",
+        description: "",
+      },
+      {
+        label: "QEMU Debug",
+        description: "",
+      },
+      {
         label: "Stop QEMU",
         description: "",
       },
@@ -81,6 +87,12 @@ export class QemuManager extends EventEmitter {
       switch (selectedOption.label) {
         case "Stop QEMU":
           QemuManager.instance.stop();
+          break;
+        case "QEMU Monitor":
+          commands.executeCommand("espIdf.monitorQemu");
+          break;
+        case "QEMU Debug":
+          commands.executeCommand("espIdf.qemuDebug");
           break;
         default:
           break;
@@ -96,10 +108,7 @@ export class QemuManager extends EventEmitter {
     idfTarget: string,
     workspaceFolder: Uri
   ) {
-    const buildPath = readParameter(
-      "idf.buildPath",
-      workspaceFolder
-    ) as string;
+    const buildPath = readParameter("idf.buildPath", workspaceFolder) as string;
     const qemuFile = Uri.joinPath(Uri.file(buildPath), "merged_qemu.bin");
     const qemuTcpPort = readParameter(
       "idf.qemuTcpPort",
@@ -137,9 +146,7 @@ export class QemuManager extends EventEmitter {
     if (this.isRunning()) {
       return;
     }
-    const modifiedEnv = await appendIdfAndToolsToPath(
-      workspaceFolder
-    );
+    const modifiedEnv = await appendIdfAndToolsToPath(workspaceFolder);
     const qemuExecutable =
       modifiedEnv.IDF_TARGET === "esp32"
         ? "qemu-system-xtensa"
@@ -148,7 +155,7 @@ export class QemuManager extends EventEmitter {
         : "";
     if (!qemuExecutable) {
       throw new Error(
-        `Current IDF_TARGET ${modifiedEnv.IDF_TARGET} is not supported in Espressif QEMU. Only esp32 or esp32c3`
+        `${modifiedEnv.IDF_TARGET} is not supported by Espressif QEMU. Only esp32 or esp32c3 targets are supported.`
       );
     }
     const isQemuBinInPath = await isBinInPath(
@@ -162,11 +169,12 @@ export class QemuManager extends EventEmitter {
       );
     }
 
-    const qemuArgs: string[] = this.getLaunchArguments(mode, modifiedEnv.IDF_TARGET, workspaceFolder);
-    if (
-      typeof qemuArgs === "undefined" ||
-      qemuArgs.length < 1
-    ) {
+    const qemuArgs: string[] = this.getLaunchArguments(
+      mode,
+      modifiedEnv.IDF_TARGET,
+      workspaceFolder
+    );
+    if (typeof qemuArgs === "undefined" || qemuArgs.length < 1) {
       throw new Error("No QEMU launch arguments found.");
     }
 
@@ -174,10 +182,7 @@ export class QemuManager extends EventEmitter {
       this.qemuTerminal = window.createTerminal({
         name: "ESP-IDF QEMU",
         env: modifiedEnv,
-        cwd:
-          workspaceFolder.fsPath ||
-          modifiedEnv.IDF_PATH ||
-          process.cwd(),
+        cwd: workspaceFolder.fsPath || modifiedEnv.IDF_PATH || process.cwd(),
         shellArgs: [],
         shellPath: env.shell,
         strictEnv: true,
@@ -220,8 +225,8 @@ export class QemuManager extends EventEmitter {
         commandDictionary[CommandKeys.QemuServer].tooltip;
       this._statusBarItem.command = CommandKeys.QemuServer;
       if (
-        commandDictionary[CommandKeys.QemuServer]
-          .checkboxState === TreeItemCheckboxState.Checked
+        commandDictionary[CommandKeys.QemuServer].checkboxState ===
+        TreeItemCheckboxState.Checked
       ) {
         this._statusBarItem.show();
       }

--- a/src/qemu/qemuManager.ts
+++ b/src/qemu/qemuManager.ts
@@ -122,7 +122,7 @@ export class QemuManager extends EventEmitter {
         "-machine",
         idfTarget,
         "-drive",
-        `file=${qemuFile.fsPath},if=mtd,format=raw`,
+        `file='${qemuFile.fsPath}',if=mtd,format=raw`,
         "-monitor stdio",
         `-serial tcp::${qemuTcpPort},server,nowait`,
       ];

--- a/templates/.devcontainer/Dockerfile
+++ b/templates/.devcontainer/Dockerfile
@@ -1,40 +1,16 @@
-FROM espressif/idf
+ARG DOCKER_TAG=latest
+FROM espressif/idf:${DOCKER_TAG}
 
-ARG DEBIAN_FRONTEND=nointeractive
 ARG CONTAINER_USER=esp
 ARG USER_UID=1050
 ARG USER_GID=$USER_UID
 
-RUN apt-get update \
-  && apt install -y -q \
-  cmake \
-  git \
-  libglib2.0-0 \
-  libnuma1 \
-  libpixman-1-0 \
-  && rm -rf /var/lib/apt/lists/*
-
-# QEMU
-ENV QEMU_REL=esp_develop_8.2.0_20240122
-ENV QEMU_SHA256=e7c72ef5705ad1444d391711088c8717fc89f42e9bf6d1487f9c2a326b8cfa83
-ENV QEMU_DIST=qemu-xtensa-softmmu-${QEMU_REL}-x86_64-linux-gnu.tar.xz
-ENV QEMU_URL=https://github.com/espressif/qemu/releases/download/esp-develop-8.2.0-20240122/${QEMU_DIST}
-
 ENV LC_ALL=C.UTF-8
 ENV LANG=C.UTF-8
-
-RUN wget --no-verbose ${QEMU_URL} \
-  && echo "${QEMU_SHA256} *${QEMU_DIST}" | sha256sum --check --strict - \
-  && tar -xf $QEMU_DIST -C /opt \
-  && rm ${QEMU_DIST}
-
-ENV PATH=/opt/qemu/bin:${PATH}
 
 RUN groupadd --gid $USER_GID $CONTAINER_USER \
   && adduser --uid $USER_UID --gid $USER_GID --disabled-password --gecos "" ${CONTAINER_USER} \
   && usermod -a -G root $CONTAINER_USER && usermod -a -G dialout $CONTAINER_USER
-
-RUN chmod -R 775 /opt/esp/python_env/
 
 USER ${CONTAINER_USER}
 ENV USER=${CONTAINER_USER}

--- a/templates/.devcontainer/Dockerfile
+++ b/templates/.devcontainer/Dockerfile
@@ -8,6 +8,8 @@ ARG USER_GID=$USER_UID
 ENV LC_ALL=C.UTF-8
 ENV LANG=C.UTF-8
 
+RUN apt-get update -y && apt-get install udev -y
+
 RUN groupadd --gid $USER_GID $CONTAINER_USER \
   && adduser --uid $USER_UID --gid $USER_GID --disabled-password --gecos "" ${CONTAINER_USER} \
   && usermod -a -G root $CONTAINER_USER && usermod -a -G dialout $CONTAINER_USER

--- a/templates/.devcontainer/Dockerfile
+++ b/templates/.devcontainer/Dockerfile
@@ -1,22 +1,10 @@
 ARG DOCKER_TAG=latest
 FROM espressif/idf:${DOCKER_TAG}
 
-ARG CONTAINER_USER=esp
-ARG USER_UID=1050
-ARG USER_GID=$USER_UID
-
 ENV LC_ALL=C.UTF-8
 ENV LANG=C.UTF-8
 
 RUN apt-get update -y && apt-get install udev -y
-
-RUN groupadd --gid $USER_GID $CONTAINER_USER \
-  && adduser --uid $USER_UID --gid $USER_GID --disabled-password --gecos "" ${CONTAINER_USER} \
-  && usermod -a -G root $CONTAINER_USER && usermod -a -G dialout $CONTAINER_USER
-
-USER ${CONTAINER_USER}
-ENV USER=${CONTAINER_USER}
-WORKDIR /home/${CONTAINER_USER}
 
 RUN echo "source /opt/esp/idf/export.sh > /dev/null 2>&1" >> ~/.bashrc
 

--- a/templates/.devcontainer/devcontainer.json
+++ b/templates/.devcontainer/devcontainer.json
@@ -15,18 +15,6 @@
 				"espressif.esp-idf-extension",
 				"espressif.esp-idf-web"
 			]
-		},
-		"codespaces": {
-			"settings": {
-				"terminal.integrated.defaultProfile.linux": "bash",
-				"idf.espIdfPath": "/opt/esp/idf",
-				"idf.toolsPath": "/opt/esp",
-				"idf.gitPath": "/usr/bin/git"
-			},
-			"extensions": [
-				"espressif.esp-idf-extension",
-				"espressif.esp-idf-web"
-			]
 		}
 	},
 	"runArgs": ["--privileged"]

